pkg/plugins/mellanox/mellanox_plugin.go

```diff
@@ -171,6 +200,97 @@ func (p *MellanoxPlugin) OnNodeStateChange(old, new *sriovnetworkv1.SriovNetwork
 	return
 }
 
+func (p *MellanoxPlugin) updateNicSpec(state *sriovnetworkv1.SriovNetworkNodeState) {
+	p.nicSyncList.Lock()
+	defer p.nicSyncList.Unlock()
+
+	var interfaces sriovnetworkv1.Interfaces
+OUTER:
+	for _, ifaceStatus := range state.Status.Interfaces {
+		if ifaceStatus.Vendor != MellanoxVendorId {
+			continue
+		}
+		for _, ifaceSpec := range state.Spec.Interfaces {
+			if ifaceSpec.PciAddress != ifaceStatus.PciAddress {
+				continue
+			}
+			// Skip non RDMA NICs
+			if !ifaceSpec.IsRdma {
+				continue OUTER
+			}
+			if ifaceSpec.NumVfs == 0 {
+				continue OUTER
+			}
+
+			interfaces = append(interfaces, ifaceSpec)
+		}
+	}
+
+	p.nicSyncList.items = interfaces
+}
+
+func (p *MellanoxPlugin) nicSpecPeriodicCheck() {
+	glog.Info("mellanox-plugin nicSpecPeriodicCheck()")
+	glog.Info("################ Start ################")
+	p.nicSyncList.RLock()
+	defer p.nicSyncList.RUnlock()
+
+	var interfaces sriovnetworkv1.Interfaces
+
+OUTER:
+	for _, iface := range p.nicSyncList.items {
+		pfLink, err := netlink.LinkByName(iface.Name)
+		if err != nil {
+			glog.Warningf("mellanox-plugin nicSpecPeriodicCheck(): failed to get PF %s Link object: %v", iface.PciAddress, err)
+			continue
+		}
+
+		for _, vf := range pfLink.Attrs().Vfs {
+			vfNetDev, err := utils.GetVFNameFromIndex(iface.PciAddress, vf.ID)
+			if err != nil {
+				glog.Warningf("mellanox-plugin nicSpecPeriodicCheck(): failed to VF NetDevName, PF pciAddress %s, VFID %d : %v", iface.PciAddress, vf.ID, err)
+				continue
+			}
+			vfRdmaDev, err := rdmamap.GetRdmaDeviceForNetdevice(vfNetDev)
+			if err != nil {
+				glog.Warningf("mellanox-plugin nicSpecPeriodicCheck(): failed to VF %s RDMA device: %v", vfNetDev, err)
+				continue
+			}
+			vfRdmaLink, err := netlink.RdmaLinkByName(vfRdmaDev)
+			if err != nil {
+				glog.Warningf("mellanox-plugin nicSpecPeriodicCheck(): failed to RDMA device %s link: %v", vfRdmaDev, err)
+				continue
+			}
+
+			trimedNodeGuid := strings.ReplaceAll(vfRdmaLink.Attrs.NodeGuid, ":", "")
+			trimedNodeGUIDMac := trimedNodeGuid[:6] + trimedNodeGuid[10:]
+
+			trimedAdminMac := strings.ReplaceAll(vf.Mac.String(), ":", "")
+
+			if trimedNodeGUIDMac != trimedAdminMac || trimedNodeGUIDMac == "000000000000" {
+				glog.Infof("mellanox-plugin nicSpecPeriodicCheck() vf info %+v", vf)
+				glog.Infof("mellanox-plugin nicSpecPeriodicCheck() trimedNodeGUIDMac %q", trimedNodeGUIDMac)
+				glog.Infof("mellanox-plugin nicSpecPeriodicCheck() trimedAdminMac %q", trimedAdminMac)
+				interfaces = append(interfaces, iface)
+				continue OUTER
+			}
+		}
+	}
+
+	if len(interfaces) == 0 {
+		glog.Info("mellanox-plugin nicSpecPeriodicCheck(): no interfaces need update")
+	}
+
+	for _, iface := range interfaces {
+		glog.Infof("mellanox-plugin nicSpecPeriodicCheck(): update node_guid and admin mac for PF %q VFs", iface.PciAddress)
+		if err := utils.SetVfsAdminMac(&iface); err != nil {
```

Comment:
As i see, the first operation is done in generic plugin. its a bit odd that the periodic is part of mellanox plugin, but i get why we are doing it in mellanox plugin.

Reply:
if there are workloads running on top of this VF, the unbind/bind happening in `utils.SetVfsAdminMac` will cause workload to fail. however seeing that in this case the device is non-healthy, i think its acceptable.

we should be careful with unbind/bind here. if sriov network device plugin runs at the wrong timing, it may miss RDMA devices.